R/testGeno.R

```diff
@@ -2,39 +2,42 @@
 ## function that gets an n\times p matrix of p genotypes of n individuals, and a null model, and tests the genotypes associations with the outcomes. 
 ## Genetic data are always assumed complete. 
 ## Types of tests: 
-## Single variant: Wald, score, BinomiRare, interaction. 
+## Single variant: Score, Score.SPA, BinomiRare, interaction. 
 ## Variant set: SKAT, burden, SKAT-O. Multiple types of p-values. Default: Davis with Koenen if does not converge. 
 
 
 # E an environmental variable for optional GxE interaction analysis. 
-testGenoSingleVar <- function(nullmod, G, E = NULL, test = c("Score", "Wald"), GxE.return.cov = FALSE){
+testGenoSingleVar <- function(nullmod, G, E = NULL, test = c("Score", "Score.SPA"),
+                              recalc.pval.thresh = 1, GxE.return.cov = FALSE){
     test <- match.arg(test)
 
     G <- .genoAsMatrix(nullmod, G)
 
-    if (test == "Wald" & nullmod$family$family != "gaussian"){
-    	test <- "Score"
-    	message("Cannot use Wald test for non-guassian families, using the score test instead.")
+    # checks on test
+    if (!is.null(E)){
+        message("Performing GxE test")
+        res <- .testGenoSingleVarWaldGxE(nullmod, G, E, GxE.return.cov.mat=GxE.return.cov)
+        return(res)
     }
-    
-    if (test == "Wald" & is.null(E)){
-        Gtilde <- calcGtilde(nullmod, G)
-        res <- .testGenoSingleVarWald(Gtilde, nullmod$Ytilde,
-                                      n=length(nullmod$Ytilde), k=ncol(nullmod$model.matrix))
+ 
+    if(test == "Score.SPA" & nullmod$family$family != "binomial"){
+        test <- "Score"
+        message("Saddlepoint approximation (SPA) can only be used for binomial family; using Score test instead.")
     }
-    
-    if (test == "Score" & !is.null(E)){
-    	test <- "Wald"
-    	message("Cannot use Score test for GxE, using the Wald test instead.")
-    }
-        
-    if (test == "Wald" & !is.null(E)){
-        res <- .testGenoSingleVarWaldGxE(nullmod, G, E, GxE.return.cov.mat=GxE.return.cov)
+
+    # run the test
+    if(test == "Score"){
+        Gtilde <- calcGtilde(nullmod, G)
+        RSS0 <- as.numeric(crossprod(nullmod$Ytilde))
+        res <- .testGenoSingleVarScore(Gtilde, G, nullmod$resid, RSS0)
```

Comment:
I had asked this before, but it might have gotten lost in the shuffle of merging branches (or I missed your response). Is there a reason not to put the Gtilde and RSS0 calculations in the `.testGenoSingleVarScore` function? The only reason I can think it would be is if passing in the whole nullmod is a bad idea.